### ISSUE TITLE
Using "void" as a type bound produces confusing errors

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#36282]
I'll note the goal of `T extends void` was intentional (to show default intent).

```dart
class G1<T> {} // OK, default.
class G2<T extends dynamic> {} // OK, same as G1.
class G3<T extends Object> {} // OK, changes resolve to bounds to Object.
class G4<T extends void> {} // ERROR: Expected a type name.
```

This occurs in the analyzer. The CFE/Dart2JS produces a different error:

```
Error compiling to JavaScript:
main.dart:1:25:
Error: Type 'void' can't be used here because it isn't a return type.
class AppView<T extends void> {}
                        ^^^^
main.dart:1:25:
Error: Type 'void' not found.
class AppView<T extends void> {}
                        ^^^^
Error: Compilation failed.
```

Both "expected a type name" and "void not found" seem wrong if this is an intentional restriction. This doesn't seem limited to one tool, so I'm adding `area-meta` so triage can help figure out what the right area(s) are.

/cc @leafpetersen

**Comment by eernstg** [comment]
I don't know if this matters here, but 'same as G1' isn't completely true:

```dart
class G1<T> {
  T t;
  void foo() { t.whatever(); } // Error.
}

class G2<T extends dynamic> {
  T t;
  void foo() { t.whatever(); } // OK.
}
```

The language team has discussed this particular kind of restrictions recently: It's specified in a quite messy manner (because `void` is a syntax error, both as a type variable bound and as a superinterface of any kind, `class C extends String {}` is prohibited in section `Strings`, etc.)

But it would definitely be nice if error messages could treat them all similarly as types.

By the way `Type 'void' can't be used here because it isn't a return type` sounds _very_ old, so that should probably never be used as an error message.

**Comment by matanlurey** [comment]
Ah, thanks for pointing that out. And yes, that error seems bizarre.

**Comment by leafpetersen** [comment]
@eernstg Is this working as intended or a spurious error?

**Comment by eernstg** [comment]
I think the error reporting is correct (that is, the part about _whether_ there is an error or not). However, the front end reports two errors for the same thing when it rejects `void` as a bound, and I suspect that whatever it is that prints "isn't a return type" it must be code written for Dart 1 (way back before 'generalized void'). Here's the output I get:
```
> cat n017.dart
// Cf. SDK issue #36282.

class G1<T> {}
class G2<T extends dynamic> {}
class G3<T extends Object> {}
class G4<T extends void> {}

main() {}
> dart n017.dart
n017.dart:6:20: Error: Type 'void' can't be used here because it isn't a return type.
Try removing 'void' keyword or replace it with 'var', 'final', or a type.
class G4<T extends void> {} // ERROR: Expected a type name.
                   ^^^^
n017.dart:6:20: Error: Type 'void' not found.
class G4<T extends void> {} // ERROR: Expected a type name.
                   ^^^^
> dartanalyzer n017.dart
Analyzing n017.dart...
  error • Expected a type name at n017.dart:6:20 • expected_type_name
1 error found.
```

**Comment by lrhn** [comment]
I'm guessing that in "it isnt' a return type", the "it" refers to the position where `void` occurs, not the `void` itself as a type. That is: "You can't use `void` here, only in a return type position" would be more correct (for Dart 1).

**Comment by eernstg** [comment]
Exactly. It sounds like that message was written in a context where `void` could actually only occur as a return type, which is a long time ago.

**Comment by srawlins** [comment]
I'm re-labeling this so its on analyzer's radar (to use a better message) and CFE's radar (to produce one error).
